### PR TITLE
Bookworm migration for sonic-mgmt-framework and sonic-gnmi dockers

### PR DESCRIPTION
**Pull request by ranjinidn**
#### Why I did it
The base OS image has already migrated to bookwork. The current sonic-mgmt-framework is still on buster and sonic-gnmi on bullseye. The changes are made to migrate the sonic-mgmt-framework and sonic-gnmi dockers to bookworm build.

##### Work item tracking
- Microsoft ADO **(number only)**:

#### How I did it
 Changed the docker manifest files to do the build for bookwork distro. The dependency packages are updated accordingly.

#### How to verify it
Built the mgmt-framework and gnmi dockers and executed the azure pipeline unit tests locally.

**Note:** The sonic-mgmt-framework build for bookworm has a dependency on the the below PR for the build to go through
https://github.com/sonic-net/sonic-mgmt-common/pull/133
https://github.com/sonic-net/sonic-mgmt-framework/pull/132

The sonic-gnmi build for bookworm has a dependency on the the below PR for the build to go through
https://github.com/sonic-net/sonic-gnmi/pull/216

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

### CODE REVIEW

File: dockers/docker-sonic-mgmt-framework/Dockerfile.j2 [hunk 1]
```diff
@@ -10,9 +10,9 @@ RUN apt-get update && \
...
 # Limited to grpcio 1.58.0 due to armhf 
 RUN pip3 install connexion==2.7.0 \
-                 setuptools==21.0.0 \
+                 --upgrade setuptools \
                  grpcio==1.58.0 \
-                 grpcio-tools==1.20.0 \
+                 grpcio-tools==1.58.0 \
                  certifi==2017.4.17 \
                  python-dateutil==2.6.0 \
                  six==1.11.0 \
```
[the remark this answers is not in the document]
Comment (amrutasali) [reply dockers/docker-sonic-mgmt-framework/Dockerfile.j2]: Addressed. Removed unnecessary packages.

**Review by kwangsuk — APPROVED** [review]

**Comment by saiarcot895** [comment]
Would you be able to handle the telemetry docker in this PR as well?

**Comment by ranjinidn** [comment]
> Would you be able to handle the telemetry docker in this PR as well?



> Would you be able to handle the telemetry docker in this PR as well?

Yes we will be able to migrate both gnmi and mgmt-framework together as sonic-mgmt-common repo is common library used by both and the bookworm migration changes are already merged to master. We are waiting on the sonic-gnmi repo changes for bookworm migration to be merged. (https://github.com/sonic-net/sonic-gnmi/pull/216)

Regarding the build failure in this PR we find a few open PRs have the same build failure reason. Hence we think the build failure is not related to the changes in this PR. How do we get past this so that this PR can be merged?

**Comment by saiarcot895** [comment]
That build failure is now fixed. Looks like there's a possibly new failure for marvell-armhf:

```
python scripts/klish_preproc_cmdtree.py /sonic/src/sonic-mgmt-framework/build/target/command-tree /sonic/src/sonic-mgmt-framework/build/target/cli-macro 6 
make[4]: python: No such file or directory
make[4]: *** [Makefile:38: /sonic/src/sonic-mgmt-framework/build/target/command-tree/.done] Error 127
make[4]: Leaving directory '/sonic/src/sonic-mgmt-framework/CLI/clitree'
make[3]: *** [Makefile:39: all] Error 2
make[3]: Leaving directory '/sonic/src/sonic-mgmt-framework/CLI'
make[2]: *** [Makefile:50: cli] Error 2
make[2]: *** Waiting for unfinished jobs....
```

This might need a change in sonic-slave-bookworm to add the python-is-python3 for all architectures, instead of just amd64 and arm64.

**Comment by ranjinidn** [comment]
@saiarcot895 we are facing some build failures and this is blocking us to test further.
The following PRs are opened to resolve the build dependencies on mgmt-framework and gnmi.
https://github.com/sonic-net/sonic-mgmt-framework/pull/141
https://github.com/sonic-net/sonic-gnmi/pull/224
However for the gnmi PR apart from build failure some pipeline Test failures are also seen. This is in discussion with the appropriate teams to resolve.  We are not sure the timeline for the above PRs to be resolved.

In the meanwhile if the submodule update for mgmt-common is reverted (PR below) we can resolve the build issue for sonic-buildimage.
https://github.com/sonic-net/sonic-buildimage/pull/18825

Please can you do the needful so we can get this PR merged for May release.

Also we tried using the python-is-python3 both in dockers/docker-sonic-mgmt-framework/Dockerfile.j2 and  sonic-slave-bookworm/Dockerfile.j2 But it was still failing for armhf with the same python error.  Any pointers will be helpful.

**Comment by amrutasali** [comment]
@saiarcot895   an update on above [comment](https://github.com/sonic-net/sonic-buildimage/pull/18548#issuecomment-2091126000)

https://github.com/sonic-net/sonic-gnmi/pull/224 has been updated to take care of pipeline test cases too.

So instead of  reverting the submodule update for mgmt-common ([PR](https://github.com/sonic-net/sonic-buildimage/pull/18825))  the build issue for sonic-buildimage can be resolved by merging the following and making submodule update for these 2 repos :
https://github.com/sonic-net/sonic-mgmt-framework/pull/141
https://github.com/sonic-net/sonic-gnmi/pull/224

**Comment by saiarcot895** [comment]
Got it, thanks @ranjinidn and @amrutasali!

Reverts of submodule updates are a bit difficult to do so given current automation, so it's better to make new commits in the submodules to fix the issue.

**Comment by saiarcot895** [comment]
As for the armhf build error, it looks like the errors are in sonic-mgmt-framework and sonic-gnmi, both of which are related to `go.mod` and `go.sum`, so I'm thinking they would get handled by the above commits.

**Comment by ranjinidn** [comment]
@saiarcot895 The builds are passing but see some failures in the kvmtests. Not very clear what is the error. Will you please help retrigger this / understand these errors.

**Comment by ranjinidn** [comment]
Thanks Sai for retriggering the tests, we still see a couple of kvmtest timeouts.
If you think this is not an issue and are random failures seen due to the setup environment, can you merge this PR?
The below 3 PRs for the individual repos are already merged for bookworm migration.
https://github.com/sonic-net/sonic-mgmt-common/pull/133
https://github.com/sonic-net/sonic-mgmt-framework/pull/132
https://github.com/sonic-net/sonic-gnmi/pull/216

**Comment by saiarcot895** [comment]
I can't approve/request a merge of this PR until all the checks are passing (unless there's a justifiable reason for the checks to fail).

It looks like the current timeouts are because it failed to lock a testbed in time, rather than any specific failure during the test execution. I've requested another rerun of those two test suites.

**Comment by ranjinidn** [comment]
@saiarcot895 we have seen the kvm tests failing due to "Acquire testbed timeout" a few times and have been rebasing our branch to master to trigger a new build. Please can you let us know if there is a way in which we can retrigger the failing cases only. Looks like we don't have permission to retrigger only the failing test.  Else each retrigger we will have to reach out to you.
It is important to push these changes for the May release as the mgmt-framwork and gnmi repos are already migrated for bookworm as mentioned earlier. Without this change the docker build for mgmt-framework and gnmi is broken currently.

File: sonic-slave-bookworm/Dockerfile.j2 [hunk 2]
```diff
@@ -694,12 +694,14 @@ RUN apt-get download python3.11-dev:$arch && apt-get download python3-dev:$arch
...
+
 # Install Bazel build system (amd64 and arm64 architectures are supported using this method)
 # TODO(PINS): Remove once pre-build Bazel binaries are available for armhf (armv7l)
 {%- if CONFIGURED_ARCH == "amd64" or CONFIGURED_ARCH == "arm64" %}
 ARG bazelisk_url=https://github.com/bazelbuild/bazelisk/releases/latest/download/bazelisk-linux-{{ CONFIGURED_ARCH }}
 RUN curl -fsSL -o /usr/local/bin/bazel ${bazelisk_url} && chmod 755 /usr/local/bin/bazel
 # Bazel requires "python"
 # TODO(PINS): remove when Bazel is okay with "python3" binary name
-RUN apt install -y python-is-python3
+#RUN apt install -y python-is-python3
```
[the remark this answers is not in the document]
Comment (amrutasali) [reply sonic-slave-bookworm/Dockerfile.j2]: Addressed

**Comment by ranjinidn** [comment]
@saiarcot895 we have a successful build for this PR. Please do merge this PR.

**Review by saiarcot895 — APPROVED** [review]

**Comment by ranjinidn** [comment]
@lguohan please review and merge this PR.

**Pull request merged**